Implement PeriodicRayBC, add multisystem support to periodic BCs

loganharbour opened the pull request:
Closes #31025

loganharbour committed:
Implement PeriodicRayBC

refs #31025

loganharbour committed:
Add more docstrings

refs #31025

loganharbour committed:
Fix typos, add param check

Refs #31025

loganharbour committed:
Remove unused declaration

refs #31025

loganharbour committed:
Set the mesh for the ghosting

Refs #31025

loganharbour committed:
Try to fix geometric ghosting

refs #31025

loganharbour committed:
Setup periodic boundary conditions for multisystem

refs #31025

loganharbour committed:
Don't query variables multiple times, cleanup AddPeriodicBCAction

loganharbour committed:
Reorganize a little, fix erase, add sanity checks

refs #31025

loganharbour committed:
Add getFieldVars()

loganharbour committed:
Add query method to translated periodic

loganharbour committed:
Use getFieldVars instead

loganharbour committed:
Reserve before filling

loganharbour committed:
Fix MooseMesh periodic bugs, more refactoring and cleanup

refs #31025

loganharbour committed:
Remove unused boundary conditions

refs #31025

loganharbour committed:
Correct spelling

refs #31025

loganharbour committed:
Update expected output with periodic ray bc change

refs #31025

loganharbour committed:
Set uninitialized values

refs #31025

loganharbour committed:
Rewrite and simplify paired sideset detection

refs #31025

loganharbour committed:
Make info messages great again

refs #31025

loganharbour committed:
Make info prettier

refs #31025

loganharbour committed:
Change error to catch what really happens

refs #31025

loganharbour committed:
Remove padding

refs #31025

loganharbour committed:
Re-order variables for padding

refs #31025

loganharbour committed:
Add error and warning checking for periodic ray bcs

refs #31280 #31025

loganharbour committed:
Add documentation for PeriodicRayBC

refs #31280 #31025

loganharbour committed:
Add documentation for SetupPeriodicRayBCAction

refs #31025

loganharbour committed:
Address conflicts after rebase

refs #31025

loganharbour committed:
Init array

refs #31025

loganharbour committed:
Simplify comment

refs #31025

loganharbour committed:
Simplify comment

refs #31025

loganharbour committed:
Use structure for unit names

refs #31025

loganharbour committed:
Communicate nonzero_dims

refs #31025

loganharbour committed:
Simplify error

refs #31025

loganharbour committed:
Don't pass problem

refs #31025

loganharbour committed:
Use common default

loganharbour committed:
Simplify error

loganharbour committed:
Gather side dimensions

refs #31025

loganharbour committed:
Require that the detection method is called first

refs #31025

loganharbour committed:
Add optimization

refs #31025

loganharbour committed:
Fix tests now that periodic BCs are not added to the aux system by default

refs #31025

loganharbour committed:
Re-add deprecated methods

refs #31025

GiudGiud commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -1646,6 +1654,12 @@ class Coupleable
   T * getVarHelper(const std::string & var_name, unsigned int comp);
 
+  /**
+   * Helper that that be used to retrieve variables of arbitrary type \p T
```
```suggestion
   * Helper that can be used to retrieve all the component variables of a variable of arbitrary type \p T
   * @param name of the variable, can be a regular, vector or array variable
```

GiudGiud commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -845,40 +856,136 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
    * @param component - An integer representing the desired component (dimension)
    */
-  bool isTranslatedPeriodic(unsigned int nonlinear_var_num, unsigned int component) const;
+  bool isTranslatedPeriodic(const MooseVariableBase & var, const unsigned int component) const;
```
this fills the same role as the other routine we don't need both

GiudGiud replied:
ok saw the use in phase_field. good with me

GiudGiud commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -845,40 +856,136 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
    * This function attempts to return the paired boundary ids for the given component.  For example,
    * in a generated 2D mesh, passing 0 for the "x" component will return (3, 1).
+   *
+   * Must call detectPairedSidesets() before calling.
```
```suggestion
   * Must have called detectPairedSidesets() before calling.
```

GiudGiud commented on framework/src/bcs/FunctionPeriodicBoundary.C:
```diff
@@ -47,58 +50,31 @@ FunctionPeriodicBoundary::get_corresponding_pos(const Point & pt) const
+    if (names.size() > i)
+    {
+      functions[i] = &problem.getFunction(names[i]);
+      const_cast<Function *>(functions[i])->initialSetup();
```
beautiful

we should really avoid calling initialSetup twice ..

GiudGiud commented on modules/ray_tracing/include/userobjects/RayTracingStudy.h:
```diff
@@ -916,6 +916,13 @@ class RayTracingStudy : public GeneralUserObject
 
+  /**
+   * Check for overlapping PeriodicRayBC boundaries and check for cases in which
+   * ghosting may not be sufficient with distribtued mesh for overlapping
```
```suggestion
   * ghosting may not be sufficient with distributed mesh for overlapping
```

GiudGiud commented on modules/ray_tracing/src/raybcs/PeriodicRayBC.C:
```diff
@@ -0,0 +1,113 @@
+  if (_periodic_neighbor == libMesh::remote_elem)
+    mooseError(
+        "Topological neighbor at ", _periodic_point, " is remote\n\n", currentRay()->getInfo());
+
```
should you assert the direction is pointing outside the boundary?
a periodic BC with functions (going from left to top for example) could leave they ray going back through the boundary the ray just got transported to

GiudGiud commented on test/tests/bcs/periodic/wedge.i:
```diff
@@ -54,6 +54,7 @@
 
   [./Periodic]
     [./x]
+      variable = u
```
```suggestion
```
shouldnt need to right

GiudGiud reviewed (changes requested):
looks great

only concern is that I don't understand the 2D/3D ray tracing results yet

loganharbour commented on framework/src/actions/AddPeriodicBCAction.C:
```diff
@@ -227,86 +82,114 @@ AddPeriodicBCAction::act()
+      mooseError("Parameter 'variable' must be specified when multiple solver systems exist");
+    // Use all field variables from system 0, excluding scalar variables
+    const auto & nl = _problem->getNonlinearSystemBase(0);
+    var_names = nl.getVariableNames();
```
technically this also supports adding adding periodicity to the aux system (which is poorly used in some of the phase field tests), so this can't be done quite yet

loganharbour committed:
Address review: improve comments

refs #31025

loganharbour committed:
Don't allow fv variables

refs #31025

loganharbour committed:
Address review: fix comments and doco

refs #31025

loganharbour committed:
Address review: don't spawn rays that have end points

This ensures that the total distance ends up being a multiple of
the ray_distance param

refs #31025

loganharbour committed:
Add testing for errors in AddPeriodicBCAction

refs #31025

loganharbour committed:
Use GhostPointNeighbors instead as the mesh ghosting functor

refs #31025

loganharbour committed:
Update tests with new ghosting

refs #31025

loganharbour committed:
Only update paired sidesets once

refs #31025
